### PR TITLE
bpo-28564: Use os.scandir() in shutil.rmtree() on Unix.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
This speeds up it to 20%.

https://bugs.python.org/issue28564

### CODE REVIEW

**Review by rhettinger — APPROVED** [review]

**Comment by pitrou** [comment]
What about the unsafe variant? Can't it benefit from a similar optimization?

**Comment by serhiy-storchaka** [comment]
Good question! Using `os.scandir()` in `_rmtree_unsafe()` speeds up it to 25% on Windows and >40% on Linux (if disable `_rmtree_safe_fd()`).

File: Lib/shutil.py [hunk 1]
```diff
@@ -394,30 +395,33 @@ def _rmtree_unsafe(path, onerror):
...
-    for name in names:
-        fullname = os.path.join(path, name)
+        onerror(os.scandir, path, sys.exc_info())
+        return
+    for entry in entries:
+        fullname = os.path.join(path, entry.name)
         try:
-            orig_st = os.stat(name, dir_fd=topfd, follow_symlinks=False)
-            mode = orig_st.st_mode
+            is_dir = entry.is_dir(follow_symlinks=False)
```
Comment (pitrou) [inline Lib/shutil.py:409]: Can you explain why you are doing this `is_dir` call before re-stat()ing below?

File: Lib/shutil.py [hunk 2]
```diff
@@ -362,25 +362,26 @@ def copytree(src, dst, symlinks=False, ignore=None, copy_function=copy2,
...
+            is_dir = False
+        if is_dir:
+            try:
+                if entry.is_symlink():
+                    # This can only happen if someone replaces
+                    # a directory with a symlink after the call to
+                    # os.scandir above.
+                    raise OSError("Cannot call rmtree on a symbolic link")
+            except OSError:
+                onerror(os.path.islink, fullname, sys.exc_info())
```
Comment (pitrou) [inline Lib/shutil.py]: Shouldn't you add a `return` just after this?

Comment (serhiy-storchaka) [reply Lib/shutil.py]: Good catch. But rather `continue`.

File: Lib/shutil.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/shutil.py:409]: If `is_dir` is false, we can avoid calling `stat()`. On Posix `scandir()` usually fills the `is_dir` bit, but `stat()` needs a syscall.

Comment (pitrou) [reply Lib/shutil.py:409]: But then why is `stat()` needed at all? Isn't `is_dir()` sufficient?

Comment (pitrou) [reply Lib/shutil.py:409]: Oh, I hadn't noticed the use of `orig_st` below. My bad.

**Review by pitrou — APPROVED** [review]
It's a nice improvement!

**Pull request merged**